sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedReader.java

```diff
@@ -299,6 +303,30 @@ public Instant getCurrentTimestamp() throws NoSuchElementException {
     return curTimestamp;
   }
 
+  @Override
+  public byte[] getCurrentRecordId() throws NoSuchElementException {
```

Comment:
can we make it an error to call this if offsetDuplication isn't set and provide a way for callers to check?

Reply:
1. Callers to check: currently callers can check via `UnboundedReader.getCurrentSource().offsetBasedDeduplicationEnabled()`. Are you proposing to provide a direct method of UnboundedReader for callers to check? For example, `UnboundedReader.offsetBasedDeduplicationEnabled()`?

2. Error to call if not set: I've adjusted the checks here a bit (check offset dedup earlier), but not sure if I captured what you meant. Please let me know if we should adjust the approach some more.

Reply:
this check is what I was thinking. See above about just generating id here

Reply:
Moved encoding, yeah this looks simpler and they should only be called once.